File: packages/analyze-trace/src/depth-limits.ts
import {
  depthLimits,
  type EventChecktypes__CheckCrossProductUnion_DepthLimit,
  type EventChecktypes__CheckTypeRelatedTo_DepthLimit,
  type EventChecktypes__GetTypeAtFlowNode_DepthLimit,
  type EventChecktypes__InstantiateType_DepthLimit,
  type EventChecktypes__RecursiveTypeRelatedTo_DepthLimit,
  type EventChecktypes__RemoveSubtypes_DepthLimit,
  type EventChecktypes__TraceUnionsOrIntersectionsTooLarge_DepthLimit,
  type EventChecktypes__TypeRelatedToDiscriminatedType_DepthLimit,
  event_checktypes__checkCrossProductUnion_DepthLimit,
  event_checktypes__checkTypeRelatedTo_DepthLimit,
  event_checktypes__getTypeAtFlowNode_DepthLimit,
  event_checktypes__instantiateType_DepthLimit,
  event_checktypes__recursiveTypeRelatedTo_DepthLimit,
  event_checktypes__removeSubtypes_DepthLimit,
  event_checktypes__traceUnionsOrIntersectionsTooLarge_DepthLimit,
  event_checktypes__typeRelatedToDiscriminatedType_DepthLimit,
  type TraceJsonSchema,
} from "@typeslayer/validate";

export type DepthLimitsRecord = {
  checkCrossProductUnion_DepthLimit: EventChecktypes__CheckCrossProductUnion_DepthLimit[];
  checkTypeRelatedTo_DepthLimit: EventChecktypes__CheckTypeRelatedTo_DepthLimit[];
  getTypeAtFlowNode_DepthLimit: EventChecktypes__GetTypeAtFlowNode_DepthLimit[];
  instantiateType_DepthLimit: EventChecktypes__InstantiateType_DepthLimit[];
  recursiveTypeRelatedTo_DepthLimit: EventChecktypes__RecursiveTypeRelatedTo_DepthLimit[];
  removeSubtypes_DepthLimit: EventChecktypes__RemoveSubtypes_DepthLimit[];
  traceUnionsOrIntersectionsTooLarge_DepthLimit: EventChecktypes__TraceUnionsOrIntersectionsTooLarge_DepthLimit[];
  typeRelatedToDiscriminatedType_DepthLimit: EventChecktypes__TypeRelatedToDiscriminatedType_DepthLimit[];
};

export const createDepthLimits = (traceFile: TraceJsonSchema) => {
  const limitNamesSet = new Set(
    depthLimits.map(d => d.shape.name),
  ) as unknown as Set<string>;
  const limitEvents = traceFile.filter(event => limitNamesSet.has(event.name));

  return {
    checkCrossProductUnion_DepthLimit: (
      limitEvents.filter(
        event =>
          event_checktypes__checkCrossProductUnion_DepthLimit.safeParse(event)
            .success,
      ) as EventChecktypes__CheckCrossProductUnion_DepthLimit[]
    ).sort((a, b) => a.args.size - b.args.size),

    checkTypeRelatedTo_DepthLimit: (
      limitEvents.filter(
        event =>
          event_checktypes__checkTypeRelatedTo_DepthLimit.safeParse(event)
            .success,
      ) as EventChecktypes__CheckTypeRelatedTo_DepthLimit[]
    ).sort((a, b) => a.args.depth - b.args.depth),

    getTypeAtFlowNode_DepthLimit: (
      limitEvents.filter(
        event =>
          event_checktypes__getTypeAtFlowNode_DepthLimit.safeParse(event)
            .success,
      ) as EventChecktypes__GetTypeAtFlowNode_DepthLimit[]
    ).sort((a, b) => a.args.flowId - b.args.flowId),

    instantiateType_DepthLimit: (
      limitEvents.filter(
        event =>
          event_checktypes__instantiateType_DepthLimit.safeParse(event).success,
      ) as EventChecktypes__InstantiateType_DepthLimit[]
    ).sort((a, b) => a.args.instantiationDepth - b.args.instantiationDepth),

    recursiveTypeRelatedTo_DepthLimit: (
      limitEvents.filter(
        event =>
          event_checktypes__recursiveTypeRelatedTo_DepthLimit.safeParse(event)
            .success,
      ) as EventChecktypes__RecursiveTypeRelatedTo_DepthLimit[]
    ).sort((a, b) => a.args.depth - b.args.depth),

    removeSubtypes_DepthLimit: limitEvents.filter(
      event =>
        event_checktypes__removeSubtypes_DepthLimit.safeParse(event).success,
    ) as EventChecktypes__RemoveSubtypes_DepthLimit[],

    traceUnionsOrIntersectionsTooLarge_DepthLimit: (
      limitEvents.filter(
        event =>
          event_checktypes__traceUnionsOrIntersectionsTooLarge_DepthLimit.safeParse(
            event,
          ).success,
      ) as EventChecktypes__TraceUnionsOrIntersectionsTooLarge_DepthLimit[]
    ).sort(
      (a, b) =>
        a.args.sourceSize * a.args.targetSize -
        b.args.sourceSize * b.args.targetSize,
    ),

    typeRelatedToDiscriminatedType_DepthLimit: (
      limitEvents.filter(
        event =>
          event_checktypes__typeRelatedToDiscriminatedType_DepthLimit.safeParse(
            event,
          ).success,
      ) as EventChecktypes__TypeRelatedToDiscriminatedType_DepthLimit[]
    ).sort((a, b) => a.args.numCombinations - b.args.numCombinations),
  } satisfies DepthLimitsRecord;
};

File: packages/typeslayer/src-tauri/src/analyze_trace/depth_limits.rs
use crate::validate::trace_json::TraceEvent;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use strum::IntoEnumIterator;
use strum_macros::EnumIter;
use strum_macros::VariantArray;
use ts_rs::TS;

#[derive(
    Eq, Hash, PartialEq, Clone, Copy, Debug, Deserialize, Serialize, EnumIter, VariantArray, TS,
)]
#[ts(export)]
pub enum DepthLimitKind {
    #[serde(rename = "checkCrossProductUnion_DepthLimit")]
    CheckCrossProductUnion,
    #[serde(rename = "checkTypeRelatedTo_DepthLimit")]
    CheckTypeRelatedTo,
    #[serde(rename = "getTypeAtFlowNode_DepthLimit")]
    GetTypeAtFlowNode,
    #[serde(rename = "instantiateType_DepthLimit")]
    InstantiateType,
    #[serde(rename = "recursiveTypeRelatedTo_DepthLimit")]
    RecursiveTypeRelatedTo,
    #[serde(rename = "removeSubtypes_DepthLimit")]
    RemoveSubtypes,
    #[serde(rename = "traceUnionsOrIntersectionsTooLarge_DepthLimit")]
    TraceUnionsOrIntersectionsTooLarge,
    #[serde(rename = "typeRelatedToDiscriminatedType_DepthLimit")]
    TypeRelatedToDiscriminatedType,
}

pub fn create_depth_limits(trace_file: &[TraceEvent]) -> HashMap<DepthLimitKind, Vec<TraceEvent>> {
    let mut depth_limits: HashMap<DepthLimitKind, Vec<TraceEvent>> = HashMap::new();

    fn kind_from_event_name(name: &str) -> Option<DepthLimitKind> {
        for kind in DepthLimitKind::iter() {
            if let Ok(serialized) = serde_plain::to_string(&kind)
                && serialized == name
            {
                return Some(kind);
            }
        }
        None
    }

    for kind in DepthLimitKind::iter() {
        depth_limits.insert(kind, Vec::new());
    }

    for ev in trace_file.iter() {
        if let Some(kind) = kind_from_event_name(ev.name())
            && let Some(vec) = depth_limits.get_mut(&kind)
        {
            vec.push(ev.clone());
        }
    }

    // Helpers to extract numeric args from specific event types
    fn num_arg(ev: &TraceEvent, key: &str) -> f64 {
        let val = serde_json::to_value(ev).ok();
        match val.and_then(|v| v.get("args").and_then(|args| args.get(key).cloned())) {
            Some(serde_json::Value::Number(n)) => n.as_f64().unwrap_or(0.0),
            _ => 0.0,
        }
    }

    // Sort per bucket based on noted criteria
    if let Some(v) = depth_limits.get_mut(&DepthLimitKind::CheckCrossProductUnion) {
        v.sort_by(|a, b| {
            num_arg(b, "size")
                .partial_cmp(&num_arg(a, "size"))
                .unwrap_or(std::cmp::Ordering::Equal)
        });
    }
    if let Some(v) = depth_limits.get_mut(&DepthLimitKind::CheckTypeRelatedTo) {
        v.sort_by(|a, b| {
            num_arg(b, "depth")
                .partial_cmp(&num_arg(a, "depth"))
                .unwrap_or(std::cmp::Ordering::Equal)
        });
    }
    if let Some(v) = depth_limits.get_mut(&DepthLimitKind::GetTypeAtFlowNode) {
        v.sort_by(|a, b| {
            num_arg(b, "flowId")
                .partial_cmp(&num_arg(a, "flowId"))
                .unwrap_or(std::cmp::Ordering::Equal)
        });
    }
    if let Some(v) = depth_limits.get_mut(&DepthLimitKind::InstantiateType) {
        v.sort_by(|a, b| {
            num_arg(b, "instantiationDepth")
                .partial_cmp(&num_arg(a, "instantiationDepth"))
                .unwrap_or(std::cmp::Ordering::Equal)
        });
    }
    if let Some(v) = depth_limits.get_mut(&DepthLimitKind::RecursiveTypeRelatedTo) {
        v.sort_by(|a, b| {
            num_arg(b, "depth")
                .partial_cmp(&num_arg(a, "depth"))
                .unwrap_or(std::cmp::Ordering::Equal)
        });
    }

    // RemoveSubtypes: not sorted

    if let Some(v) = depth_limits.get_mut(&DepthLimitKind::TraceUnionsOrIntersectionsTooLarge) {
        v.sort_by(|a, b| {
            let a_prod = num_arg(a, "sourceSize") * num_arg(a, "targetSize");
            let b_prod = num_arg(b, "sourceSize") * num_arg(b, "targetSize");
            b_prod
                .partial_cmp(&a_prod)
                .unwrap_or(std::cmp::Ordering::Equal)
        });
    }
    if let Some(v) = depth_limits.get_mut(&DepthLimitKind::TypeRelatedToDiscriminatedType) {
        v.sort_by(|a, b| {
            num_arg(b, "numCombinations")
                .partial_cmp(&num_arg(a, "numCombinations"))
                .unwrap_or(std::cmp::Ordering::Equal)
        });
    }

    depth_limits
}
